flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkRowData.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink;
+
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.logical.LogicalType;
+
+public class FlinkRowData {
+
+  private FlinkRowData() {}
+
+  public static RowData.FieldGetter createFieldGetter(LogicalType fieldType, int fieldPos) {
+    RowData.FieldGetter flinkFieldGetter = RowData.createFieldGetter(fieldType, fieldPos);
+    return rowData -> {
+      // Be sure to check for null values, even if the field is required. Flink
+      // RowData.createFieldGetter(..) does not null-check optional / nullable types. Without this
+      // explicit null check, the null flag of BinaryRowData will be ignored and random bytes will
+      // be parsed as actual values. This will produce incorrect writes instead of failing with a
+      // NullPointerException.
+      if (!fieldType.isNullable() && rowData.isNullAt(fieldPos)) {
+        return null;
+      }
+      return flinkFieldGetter.getFieldOrNull(rowData);
+    };
+  }
```

Comment:
Moved the core change here to replace all the `RowData#createFieldGetter` calls. The idea is to also perform a null check for non-null types to prevent interpreting nulled fields in BinaryRowData as actual values. Unfortunately, Flink itself only adds the null check for nullable types and defers additional null checks to the caller. I'll report this in upstream Flink as well.

Reply:
I like this new approach. 

I am also wondering if this is also a bug in Flink where `RowData.createFieldGetter` couldn't handle the `BinaryRowData` properly regarding null value.

Reply:
Yes, I think it could be seen as a bug. Filed a ticket here: https://issues.apache.org/jira/browse/FLINK-37245

Reply:
please add the JIRA issue to the comments. We can clean this up after the upstream issue is resolved.

Reply:
Done.